ethcore/src/spec/spec.rs

```diff
@@ -540,7 +526,18 @@ impl Spec {
 
 		// TODO: could optimize so we don't re-run, but `ensure_db_good` is barely ever
 		// called anyway.
-		let db = self.run_constructors(factories, db)?;
+		let (root, db) = run_constructors(
+			&self.genesis_state,
+			&self.constructors,
+			&*self.engine,
+			self.author,
+			self.timestamp,
+			self.difficulty,
+			factories,
+			db
+		)?;
+
+		assert_eq!(root, self.state_root(), "Spec's state root has not been precomputed correctly.");
```

Comment:
Is it safe / expected to panic here?

Reply:
yes, cause `self.state_root()` is always created during init with the same constructors